File: taurus-protect-sdk-typescript/src/mappers/user-device.ts
/**
 * User device mapper functions for converting OpenAPI DTOs to domain models.
 */

import type {
  UserDevicePairing,
  UserDevicePairingInfo,
  UserDevicePairingStatus,
} from '../models/user-device';
import { safeString } from './base';

/**
 * Maps a CreateUserDevicePairingReply DTO to a UserDevicePairing domain model.
 *
 * @param dto - The DTO from the OpenAPI client
 * @returns The domain model
 */
export function userDevicePairingFromDto(dto: unknown): UserDevicePairing | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  const pairingId = safeString(d.pairingID ?? d.pairingId ?? d.pairing_id);

  if (!pairingId) {
    return undefined;
  }

  return {
    pairingId,
  };
}

/**
 * Maps a UserDevicePairingInfo DTO to a UserDevicePairingInfo domain model.
 *
 * @param dto - The DTO from the OpenAPI client
 * @returns The domain model
 */
export function userDevicePairingInfoFromDto(dto: unknown): UserDevicePairingInfo | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  const pairingId = safeString(d.pairingID ?? d.pairingId ?? d.pairing_id);
  const statusValue = d.status;

  if (!pairingId || statusValue === undefined) {
    return undefined;
  }

  // Extract status value - handle both string and object formats
  let status: UserDevicePairingStatus;
  if (typeof statusValue === 'string') {
    status = statusValue as UserDevicePairingStatus;
  } else if (typeof statusValue === 'object' && statusValue !== null) {
    // Handle case where status might be an enum object
    status = String(statusValue) as UserDevicePairingStatus;
  } else {
    return undefined;
  }

  return {
    pairingId,
    status,
    apiKey: safeString(d.apiKey ?? d.api_key),
  };
}

File: taurus-protect-sdk-python/taurus_protect/mappers/user_device.py
"""User device mapper for converting OpenAPI DTOs to domain models."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Optional

from taurus_protect.mappers._base import safe_datetime, safe_string
from taurus_protect.models.user_device import UserDevicePairing, UserDevicePairingInfo

if TYPE_CHECKING:
    pass  # For OpenAPI types when available


def user_device_pairing_from_dto(dto: Any) -> Optional[UserDevicePairing]:
    """
    Convert OpenAPI TgvalidatordCreateUserDevicePairingReply to domain UserDevicePairing.

    Args:
        dto: OpenAPI user device pairing reply DTO.

    Returns:
        Domain UserDevicePairing model or None if dto is None.
    """
    if dto is None:
        return None

    return UserDevicePairing(
        pairing_id=safe_string(getattr(dto, "pairing_id", None)),
        status=safe_string(getattr(dto, "status", None)),
        created_at=safe_datetime(getattr(dto, "creation_date", None)),
        expires_at=safe_datetime(getattr(dto, "expiration_date", None)),
    )


def user_device_pairing_info_from_dto(dto: Any) -> Optional[UserDevicePairingInfo]:
    """
    Convert OpenAPI TgvalidatordUserDevicePairingInfo to domain UserDevicePairingInfo.

    Args:
        dto: OpenAPI user device pairing info DTO.

    Returns:
        Domain UserDevicePairingInfo model or None if dto is None.
    """
    if dto is None:
        return None

    return UserDevicePairingInfo(
        pairing_id=safe_string(getattr(dto, "pairing_id", None)),
        user_id=safe_string(getattr(dto, "user_id", None)),
        status=safe_string(getattr(dto, "status", None)),
        device_name=safe_string(getattr(dto, "device_name", None)),
        device_type=safe_string(getattr(dto, "device_type", None)),
        encryption_key=safe_string(getattr(dto, "encryption_key", None)),
        created_at=safe_datetime(getattr(dto, "creation_date", None)),
        expires_at=safe_datetime(getattr(dto, "expiration_date", None)),
    )
